datafusion/core/src/physical_optimizer/repartition.rs

```diff
@@ -163,10 +178,34 @@ fn optimize_partitions(
             .children()
             .iter()
             .map(|child| {
+                // does plan itelf (not parent) require its input to
+                // be sorted in some way?
+                let required_input_ordering =
+                    plan_has_required_input_ordering(plan.as_ref());
+
+                let can_reorder_child = if can_reorder {
+                    // parent of `plan` will not use any particular order
+
+                    // if `plan` itself doesn't need order OR
+                    !required_input_ordering ||
+                    // child has no order to preserve
+                        child.output_ordering().is_none()
+                } else {
+                    // parent would like to use the `plan`'s output
+                    // order.
+
+                    // if `plan` doesn't maintain the input order and
+                    // doesn't need the child's output order itself
+                    (!plan.maintains_input_order() &&  !required_input_ordering) ||
+                    // child has no ordering to preserve
+                        child.output_ordering().is_none()
+                };
```

Comment:
This is the logic I eventually arrived at after a lot of thought. I tried to document the rationale as best I could -- perhaps @ozankabak  and @mustafasrepo  might have a few minutes to double check this. I believe this was first introduced by https://github.com/apache/arrow-datafusion/pull/4691

Reply:
Sure; will take a look, discuss with the team and get back to you.

Reply:
@alamb I have checked the snippet. I have tried a couple of alternatives with no success. I think this behavior is correct.

Reply:
Looks good to me as well